rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,322 @@
+    marker: PhantomData<&'pr yp_diagnostic_t>
+}
+
+impl<'pr> Iterator for Diagnostics<'pr> {
+    type Item = Diagnostic<'pr>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.diagnostic.is_null() {
+            None
+        } else {
+            let current = Diagnostic {
+                diagnostic: unsafe { NonNull::new_unchecked(self.diagnostic) },
+                marker: PhantomData
+            };
+
+            self.diagnostic = unsafe { (*(self.diagnostic)).node.next.cast::<yp_diagnostic_t>() };
+            Some(current)
+        }
+    }
+}
+
+/// A struct created by the `comments` method on `ParseResult`. It can be used
+/// to iterate over the comments in the parse result.
+pub struct Comments<'pr> {
+    comment: *mut yp_comment_t,
+    marker: PhantomData<&'pr yp_comment_t>
+}
+
+impl<'pr> Iterator for Comments<'pr> {
+    type Item = Comment<'pr>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.comment.is_null() {
+            None
+        } else {
+            let current = Comment {
+                comment: unsafe { NonNull::new_unchecked(self.comment) },
+                marker: PhantomData
+            };
+
+            self.comment = unsafe { (*(self.comment)).node.next.cast::<yp_comment_t>() };
+            Some(current)
+        }
+    }
+}
+
+/// The result of parsing a source string.
+pub struct ParseResult<'pr> {
+    source: &'pr [u8],
+    parser: NonNull<yp_parser_t>,
+    node: NonNull<yp_node_t>
+}
+
+impl<'pr> ParseResult<'pr> {
+    /// Returns the source string that was parsed.
+    #[must_use]
+    pub const fn source(&self) -> &'pr [u8] {
+        self.source
+    }
+
+    /// Returns a slice of the source string that was parsed using the given
+    /// location range.
+    #[must_use]
+    pub fn slice(&self, location: &Location<'pr>) -> &'pr [u8] {
+        let ptr = self.source.as_ptr() as usize;
+        &self.source[(location.start() - ptr)..(location.end() - ptr)]
+    }
+
+    /// Returns an iterator that can be used to iterate over the errors in the
+    /// parse result.
+    #[must_use]
+    pub fn errors(&self) -> Diagnostics<'_> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).error_list;
+            Diagnostics { diagnostic: list.head.cast::<yp_diagnostic_t>(), marker: PhantomData }
+        }
+    }
+
+    /// Returns an iterator that can be used to iterate over the warnings in the
+    /// parse result.
+    #[must_use]
+    pub fn warnings(&self) -> Diagnostics<'_> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).warning_list;
+            Diagnostics { diagnostic: list.head.cast::<yp_diagnostic_t>(), marker: PhantomData }
+        }
+    }
+
+    /// Returns an iterator that can be used to iterate over the comments in the
+    /// parse result.
+    #[must_use]
+    pub fn comments(&self) -> Comments<'_> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).comment_list;
```

Comment:
If `comment_list` is a raw pointer, I would avoid making a Rust reference out of it.

Reply:
In this case `comment_list` is embedded directly onto the parser. So it's not a pointer, it's the struct itself. I want to get a pointer to it, though, so that I can avoid copying it.